File: flutter_pos_2013/lib/core/utils/hmac_util.dart
import 'dart:convert';
import 'dart:math';
import 'dart:typed_data';
import 'package:crypto/crypto.dart';
import 'package:uuid/uuid.dart';

/// HMAC-SHA256 Signature Generator for Protocol 201.3
class HmacUtil {
  static const String algorithm = 'HmacSHA256';
  static const String _keyAlias = 'POS_HMAC_KEY';
  
  /// Generate HMAC-SHA256 signature for batch upload
  static String generateSignature({
    required String protocolVersion,
    required String merchantId,
    required String terminalId,
    required String batchId,
    required int timestamp,
    required String nonce,
    int transactionCount = 1,
    required String secretKey,
  }) {
    final payload = '$protocolVersion|$merchantId|$terminalId|$batchId|$timestamp|$nonce|$transactionCount';
    return generateHmac(payload, secretKey);
  }

  /// Generate HMAC-SHA256 for a string payload
  static String generateHmac(String payload, String secretKey) {
    final key = utf8.encode(secretKey);
    final bytes = utf8.encode(payload);
    
    final hmacSha256 = Hmac(sha256);
    final digest = hmacSha256.convert(bytes);
    
    return base64Encode(digest.bytes);
  }

  /// Generate a random nonce
  static String generateNonce() {
    return const Uuid().v4().replaceAll('-', '').substring(0, 16);
  }

  /// Generate a batch ID
  static String generateBatchId() {
    final timestamp = DateTime.now().millisecondsSinceEpoch;
    final random = Random().nextInt(9999).toString().padLeft(4, '0');
    return 'BATCH-$timestamp-$random';
  }

  /// Verify a signature
  static bool verifySignature({
    required String signature,
    required String protocolVersion,
    required String merchantId,
    required String terminalId,
    required String batchId,
    required int timestamp,
    required String nonce,
    int transactionCount = 1,
    required String secretKey,
  }) {
    final expected = generateSignature(
      protocolVersion: protocolVersion,
      merchantId: merchantId,
      terminalId: terminalId,
      batchId: batchId,
      timestamp: timestamp,
      nonce: nonce,
      transactionCount: transactionCount,
      secretKey: secretKey,
    );
    return signature == expected;
  }

  /// Hash data with SHA-256
  static String sha256Hash(String data) {
    final bytes = utf8.encode(data);
    final digest = sha256.convert(bytes);
    return base64Encode(digest.bytes);
  }

  /// Generate a secure random string
  static String generateSecureRandom(int length) {
    final random = Random.secure();
    final values = List<int>.generate(length, (i) => random.nextInt(256));
    return base64UrlEncode(values).substring(0, length);
  }
}

File: android_pos_app/app/src/main/java/com/pos2013/offline/utils/HmacUtil.kt
package com.pos2013.offline.utils

import android.security.keystore.KeyGenParameterSpec
import android.security.keystore.KeyProperties
import android.util.Base64
import com.pos2013.offline.config.GatewayConfig
import java.nio.charset.StandardCharsets
import java.security.KeyStore
import javax.crypto.Mac
import javax.crypto.SecretKey
import javax.crypto.spec.SecretKeySpec
import javax.crypto.KeyGenerator

/**
 * HMAC-SHA256 signature generator for Protocol 201.3
 */
object HmacUtil {

    private const val ALGORITHM = "HmacSHA256"
    private const val ANDROID_KEYSTORE = "AndroidKeyStore"
    private const val KEY_ALIAS = "POS_HMAC_KEY"

    /**
     * Generate HMAC-SHA256 signature for batch upload
     */
    fun generateSignature(
        protocolVersion: String,
        merchantId: String,
        terminalId: String,
        batchId: String,
        timestamp: Long,
        nonce: String,
        transactionCount: Int = 1,
        secretKey: String = GatewayConfig.GATEWAY_SECRET_KEY
    ): String {
        val payload = "$protocolVersion|$merchantId|$terminalId|$batchId|$timestamp|$nonce|$transactionCount"
        return generateHmac(payload, secretKey)
    }

    /**
     * Generate HMAC-SHA256 for a string payload.
     * Accessible within the package or as public for repositories.
     */
    fun generateHmac(payload: String, secretKey: String): String {
        try {
            val keyStore = KeyStore.getInstance(ANDROID_KEYSTORE)
            keyStore.load(null)

            if (!keyStore.containsAlias(KEY_ALIAS)) {
                try {
                    val keyGenerator = KeyGenerator.getInstance(ALGORITHM, ANDROID_KEYSTORE)
                    val spec = KeyGenParameterSpec.Builder(
                        KEY_ALIAS,
                        KeyProperties.PURPOSE_SIGN or KeyProperties.PURPOSE_VERIFY
                    ).setDigests(KeyProperties.DIGEST_SHA256)
                        .setKeySize(256)
                        .setUserAuthenticationRequired(false)
                        .build()

                    keyGenerator.init(spec)
                    keyGenerator.generateKey()
                } catch (genEx: Exception) {
                }
            }

            val entry = keyStore.getEntry(KEY_ALIAS, null)
            if (entry is KeyStore.SecretKeyEntry) {
                val secret = entry.secretKey
                val mac = Mac.getInstance(ALGORITHM)
                mac.init(secret)
                val signatureBytes = mac.doFinal(payload.toByteArray(StandardCharsets.UTF_8))
                return Base64.encodeToString(signatureBytes, Base64.NO_WRAP)
            }
        } catch (e: Exception) {
        }

        val mac = Mac.getInstance(ALGORITHM)
        val secretKeySpec = SecretKeySpec(secretKey.toByteArray(StandardCharsets.UTF_8), ALGORITHM)
        mac.init(secretKeySpec)

        val signatureBytes = mac.doFinal(payload.toByteArray(StandardCharsets.UTF_8))
        return Base64.encodeToString(signatureBytes, Base64.NO_WRAP)
    }

    fun generateNonce(): String {
        return java.util.UUID.randomUUID().toString().replace("-", "").take(16)
    }
}
